corehq/apps/app_manager/util.py

```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
     return sort_only_elements, sort_columns
 
 
-def get_form_data(domain, app, include_shadow_forms=True):
-    from corehq.apps.reports.formdetails.readable import FormQuestionResponse
-    from corehq.apps.app_manager.models import ShadowForm
-
-    case_meta = app.get_case_metadata()
-    modules = []
-    errors = []
-    for module in app.get_modules():
-        forms = []
-        module_meta = {
+class FormDataGenerator(object):
+    def __init__(self, domain, app, include_shadow_forms=True):
+        self.domain = domain
+        self.app = app
+        self.include_shadow_forms = include_shadow_forms
+
+        self.errors = []
+
+        self._seen_save_to_case = defaultdict(list)
+        self._case_meta = self.app.get_case_metadata()
+
+    def get_app_data(self):
+        return [self._compile_module(module) for module in self.app.get_modules()], self.errors
+
+    def _compile_module(self, module):
+        return {
             'id': module.unique_id,
             'name': module.name,
             'short_comment': module.short_comment,
             'module_type': module.module_type,
             'is_surveys': module.is_surveys,
             'module_filter': module.module_filter,
+            'forms': [self._compile_form(form) for form in self._get_pertinent_forms(module)],
         }
 
-        form_list = module.get_forms()
-        if not include_shadow_forms:
-            form_list = [f for f in form_list if not isinstance(f, ShadowForm)]
-        for form in form_list:
-            form_meta = {
-                'id': form.unique_id,
-                'name': form.name,
-                'short_comment': form.short_comment,
-                'action_type': form.get_action_type(),
-                'form_filter': form.form_filter,
-                'questions': [],
+    def _get_pertinent_forms(self, module):
+        from corehq.apps.app_manager.models import ShadowForm
+        if not self.include_shadow_forms:
+            return [form for form in module.get_forms() if not isinstance(form, ShadowForm)]
+        return module.get_forms()
+
+    def _compile_form(self, form):
+        form_meta = {
+            'id': form.unique_id,
+            'name': form.name,
+            'short_comment': form.short_comment,
+            'action_type': form.get_action_type(),
+            'form_filter': form.form_filter,
+        }
+        try:
+            form_meta['questions'] = [
+                compiled_question
+                for question in form.get_questions(self.app.langs, include_triggers=True,
+                                                   include_groups=True, include_translations=True)
+                for compiled_question in self._compile_form_questions(form.unique_id, question)
```

Comment:
`self._compile_form_questions()` always yields a single item. What do you think of changing it to `return` rather than `yield`, and rephrase this as
```py
form_meta['questions'] = [
    self._compile_form_question(form.unique_id, question)
    for question in form.get_questions(
        self.app.langs, include_triggers=True,
        include_groups=True, include_translations=True)
]
```
?

Reply:
Thanks for this. It actually is supposed to yield 2 items if the question requires a new save to case block to be added. I've updated that in a later commit.